src/query/ts/m3db/encoded_block_iterator.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"time"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/block"
+	"github.com/m3db/m3/src/query/functions/utils"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage"
+)
+
+type consolidationSettings struct {
+	consolidationFn ConsolidationFunc
+	currentTime     time.Time
+	bounds          *models.Bounds
+}
+
+type encodedBlock struct {
+	lastBlock            bool
+	meta                 block.Metadata
+	tagOptions           models.TagOptions
+	consolidation        *consolidationSettings
+	seriesMetas          []block.SeriesMeta
+	seriesBlockIterators []encoding.SeriesIterator
+}
+
+// verify encodedSeriesIter satisfies both block and unconsolidated block
+var (
+	_ block.Block               = (*encodedBlock)(nil)
```

Comment:
rather have MakeEncodedBlock return block.Block. Also, not sure about it being both a block and unconsolidatedblock

Reply:
Refactored to do it this way

Reply:
Longer discussion, but are you guys familiar with the "accept interfaces return structs" dictum? Not the pattern in this repo, but it is elsewhere within Uber. https://medium.com/@cep21/what-accept-interfaces-return-structs-means-in-go-2fe879e25ee8

Reply:
Also mentioned in [the go libraries talk](https://docs.google.com/presentation/d/1phfzoGQqEpSQVXRiy0Zpq5v2USy9bc6Bc2XRrV0R9sA/edit)

Reply:
hmm interesting, will take a look